mapbox/services/directions.py

```diff
@@ -1,103 +1,223 @@
+        result = []
+        if bearings is None:
+            return None
+        if len(bearings) != len(features):
+            raise errors.InvalidParameterError(
+                'Must provide exactly one bearing tuple for each input feature')
+        for bearing in bearings:
+            if bearing is None:
+                result.append(None)
+            else:
+                try:
+                    angle, rng = bearing
+                except ValueError:
+                    raise errors.InvalidParameterError(
+                        'bearing tuple must contain two elements: (angle, range)')
+
+                if angle <= 0 or angle >= 360 or rng <= 0 or rng >= 360:
+                    raise errors.InvalidParameterError(
+                        'angle and range must be between 0 and 360')
+                result.append((angle, rng))
+        return result
+
+    def _validate_radiuses(self, radiuses, features):
+        result = []
+        if radiuses is None:
+            return None
+        if len(radiuses) != len(features):
+            raise errors.InvalidParameterError(
+                'Must provide exactly one radius for each input feature')
+        for radius in radiuses:
+            if radius == 'unlimited' or radius > 0:
+                result.append(radius)
+            else:
+                raise errors.InvalidParameterError(
+                    '{0} is not a valid radius'.format(radius))
+        return result
+
+    @staticmethod
+    def _encode_bearing(b):
+        if b is None:
+            return ''
+        else:
+            return '{},{}'.format(*b)
+
+    def directions(self, features, profile='mapbox/driving', alternatives=None,
+                   geometries=None, overview=None, radiuses=None, steps=None,
+                   continue_straight=None, bearings=None, annotations=None,
+                   language=None, **kwargs):
         """Request directions for waypoints encoded as GeoJSON features.
 
-        :param features: sequence of GeoJSON features.
-        :param profile: name of a profile.
+        Parameters
+        ----------
+        features: iterable of GeoJSON-like Feature mappings
+        profile: string
+        alternatives: boolean
+        geometries: string
+            Type of geometry returned (geojson, polyline, polyline6)
+        overview: string or False
+            ('full', 'simplified', False)
+        radiuses: iterable of numbers or 'unlimited'
+            Must be same length as features
+        steps: boolean
+        continue_straight: boolean
+        bearings: list of bearing 2-tuples
+            where the bearing tuple consists of an angle and range
+                [(215, 45), (315, 90)]
+            Must be same length as features
+        annotations: string
+        language: string
+
+        Returns
+        -------
+        requests.Response
+            the json() method will return a Directions response dict
```

Comment:
I added a combined `waypoint_snapping` param in the latest commit (f7e91e9)

The parameter is a bit loose with types and there are a number of potential combos of inputs. From the docstring:

```
waypoint_snapping: list
    List must be same length as features.
    Elements of the list must be one of:
    - A number (interpretted as a snapping radius)
    - The string 'unlimited' (unlimited snapping radius)
    - A 3-element tuple consisting of (radius, angle, range)
    - None (no snapping parameters specified for that waypoint)
```

Makes internal validation code a little complex but IMO improves usability.